tool/dart_model_generator/lib/generate_dart_model.dart

```diff
@@ -0,0 +1,136 @@
+
+import 'dart:io';
+
+import 'package:dart_style/dart_style.dart';
+import 'package:json_schema/json_schema.dart';
+
+class DartModelGenerator {
+  void run() {
+    final result = <String>[];
+    final schema = JsonSchema.create(
+        File('schemas/dart_model.schema.json').readAsStringSync());
+    for (final def in schema.defs.entries) {
+      result.add(_generateExtensionType(def.key, def.value));
+    }
+
+    File('pkgs/dart_model/lib/src/dart_model.g.dart').writeAsStringSync(
+        DartFormatter().formatSource(SourceCode(result.join('\n'))).text);
+  }
+
+  String _generateExtensionType(String name, JsonSchema definition) {
+    final result = StringBuffer();
+
+    final jsonType = switch (definition.type) {
+      SchemaType.object => 'Map<String, Object?> node',
+      SchemaType.string => 'String string',
+      _ => throw UnsupportedError('Schema type ${definition.type}.'),
+    };
+    result.writeln('extension type $name.fromJson($jsonType) {');
+
+    final propertyMetadatas = definition.properties.entries
+        .map((e) => _readPropertyMetadata(e.key, e.value))
+        .toList();
+    if (definition.type == SchemaType.object) {
+      result.writeln('  $name({');
+      for (final property in propertyMetadatas) {
+        result.writeln(switch (property.type) {
+          PropertyType.object =>
+            '${property.elementTypeName}? ${property.name},',
+          PropertyType.bool => 'bool? ${property.name},',
+          PropertyType.string => 'string? ${property.name},',
+          PropertyType.list =>
+            'List<${property.elementTypeName}>? ${property.name},',
+          PropertyType.map =>
+            'Map<String, ${property.elementTypeName}>? ${property.name},',
+        });
+      }
+      result.writeln('}) : this.fromJson({');
+      for (final property in propertyMetadatas) {
+        result.writeln('if (${property.name} != null) '
+            "'${property.name}': ${property.name},");
+      }
+      result.writeln('});');
+    } else if (definition.type == SchemaType.string) {
+      result.writeln('$name(String string) : this.fromJson(string);');
+    }
+
+    for (final property in propertyMetadatas) {
+      result.writeln(switch (property.type) {
+        PropertyType.object =>
+          // TODO(davidmorgan): use the extension type constructor instead of
+          // casting.
+          '${property.elementTypeName} get ${property.name} => '
+              'node[\'${property.name}\']!'
+              ' as ${property.elementTypeName};',
+        PropertyType.bool => 'bool get ${property.name} => '
+            'node[\'${property.name}\']! as bool;',
+        PropertyType.string => 'String get ${property.name} => '
+            'node[\'${property.name}\']! as String;',
+        PropertyType.list =>
+          'List<${property.elementTypeName}> get ${property.name} => '
+              'node[\'${property.name}\']! '
+              'as List<${property.elementTypeName}>;',
```

Comment:
Can `property.elementTypeName` be `String`? If so, I think this cast will fail (it will be a `List<Object?>`).

Actually, I don't think this can work at all since the extensions are on a `Map<String, Object?>`? And this is a `List<Object?>` presumably?

Reply:
Added coverage to `model_test.dart`, by adding a metadata annotation, "field" accesses, and serialization. It works with the data created in the test but not after deserialization :)

Good catch, thanks! Changed to use "cast()".